walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,120 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> availableTimeRanges = new ArrayList<>();
+    Set<String> attendees = new HashSet<>(request.getAttendees());
+    long duration = request.getDuration();
+
+    if (Long.compare(duration, TimeRange.WHOLE_DAY.duration()) > 0) {
+      // If the requested duration is longer than the {@code WHOLE_DAY}, there can't be any
+      // {@code TimeRange} that satisfies the {@code request}.
+      return Arrays.asList();
+    } else if (events.isEmpty() || attendees.isEmpty()) {
+      // If there are no existing {@code Event} or mandatory attendees, the {@code WHOLE_DAY}
+      // satisfies the request.
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    // Find existing {@code TimeRange} that correspond to {@code Event} having overlapping
+    // attendees with the {@code request}.
+    Collection<TimeRange> relevantTimeRanges = new HashSet<>();
+    for (Event event : events) {
+      // Make a shallow copy of the original set of attendees.
+      Set<String> overlappingAttendees = new HashSet<>(event.getAttendees());
+      overlappingAttendees.retainAll(attendees);
+      if (!overlappingAttendees.isEmpty()) {
+        relevantTimeRanges.add(event.getWhen());
+      }
+    }
+    if (relevantTimeRanges.isEmpty()) {
+      // If there are no existing {@code Event} that contain overlapping attendees with the meeting
```

Comment:
Remove this comment. Instead of explaining **what** your code does, try to explain **why** it does what it does. Another programmer reading your code can figure out what it does, but it's not so obvious why it is behaving this way. Try to make the function assumptions, contracts etc. more explicit in the function description itself.

Reply:
Got it, thank you!